pages/quaternions.md

```diff
@@ -0,0 +1,51 @@
+# Quaternions
+[Quaternions](https://en.wikipedia.org/wiki/Quaternion) are best known for their suitability as representations of 3D rotational orientation.
+They can also be viewed as an extension of complex numbers.
+
+Many softwares such as [SciPy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.html) and [ROS](https://wiki.ros.org/tf2/Tutorials/Quaternions), they treat the order of a quaternion as $xi+yj+zk+w$, but not $w+xi+yj+zk$.
+One reason for that is that the order in [swizzling](https://en.wikipedia.org/wiki/Swizzling_(computer_graphics)) is `xyzw`.
+
+However, most Julia packages use $w+xi+yj+zk$ order.
+This is for consistency with `Base.Complex`'s $x+iy$ order.
+
+\toc
+
+## Packages for Quaternions
+
+### Quaternions.jl
+[Quaternions.jl](https://github.com/JuliaGeometry/Quaternions.jl) is the most popular Julia package for quaternions.
+
+* This package implements `Quaternions.Quaternion` which is much similar to `Base.Complex`.
+* This package focus on some basic operations for quaternions. Most implemented methods are added to `Base` functions.
+* Other operations such as rotations are not implemented in this package. Use [Rotations.jl](https://github.com/JuliaGeometry/Rotations.jl) instead.
+* The arguments for `Quaternions.Quaternion` is ordered in $w+xi+yj+zk$.
+
+Historically, this package was not actively maintained before 2022.
+This causes the following confusion.
```

Comment:
```suggestion
```

It didn't help that `Quaternions` wasn't actively maintained, but that's not really the reason `Quaternionic` exists.  They really are different packages, and I feel like this line implies that the more active maintenance now happening makes `Quaternionic` unnecessary, when really there's a lot more in `Quaternionic` than is in `Quaternions`.

Reply:
I see. Then, what would be the most bad part about Quaternions.jl?

>In particular, [the most popular](https://github.com/JuliaGeometry/Quaternions.jl) of those has made certain design choices that conflict with my needs.

I thought this text in [Quaternionic.jl README](https://github.com/moble/Quaternionic.jl) was related to these issues:

* `Quaternions.Quaternion` had incomplete compatibility with `Base.Complex`. https://github.com/JuliaGeometry/Quaternions.jl/issues/62
* `Quaternions.Quaternion` had `norm::Bool` field. https://github.com/JuliaGeometry/Quaternions.jl/issues/60.

Reply:
Well, first I should say that I don't consider most of the differences "bad"; they're just incompatible with my needs.  Different people make different decisions about what should go in a package.  But:

1. I don't think the first issue you point out was exactly what I was worried about, but there are indeed several places where `Quaternions` assumes that ambiguity in the direction of the vector part should resolve in the `x` direction.  This conflicts with common usage in physics and the theory of special functions, where the `z` direction is the most natural choice.
2. The second issue you point out was certainly part of my biggest concern: the lack of type structure.  Having `Quaternion`, `Rotor`, and `QuatVec` really helps ensure that operations make sense and that methods can be implemented in the most accurate and efficient ways.

Also, I just generally felt that the `Quaternions` devs were pushing to keep it a small package, whereas there are a lot more things that I wanted quaternions to be able to do.

Reply:
Thank you for the comment!

>Also, I just generally felt that the Quaternions devs were pushing to keep it a small package, whereas there are a lot more things that I wanted quaternions to be able to do.

I also have the same thoughts. I'm also a maintainer of [Rotations.jl](https://github.com/JuliaGeometry/Rotations.jl), and I thought keeping Quaternions.jl small would help other packages that depend on Quaternions.jl.

>the lack of type structure. Having `Quaternion`, `Rotor`, and `QuatVec` really helps ensure that operations make sense and that methods can be implemented in the most accurate and efficient ways.

There was a related discussion about having `Quaternions.AbstractQuaternion` (https://github.com/JuliaGeometry/Quaternions.jl/issues/79#issuecomment-1059742924). I was thinking that we don't need that because we don't have `Base.AbstractComplex` (https://github.com/JuliaLang/julia/issues/33246).

But I now feel it would be okay to have `Quaternions.AbstractQuaternion` because https://github.com/JuliaLang/julia/pull/35587 is almost ready to merge..

Reply:
Can you also remove the "This causes the following confusion." line, please?  After that, I'm satisfied.